src/ui/components/shared/WorkspaceSettingsModal/WorkspaceMember.tsx

```diff
@@ -70,52 +80,64 @@ function WorkspaceMemberRoles({
   member: WorkspaceUser;
   onClick: () => void;
 }) {
-  const role: WorkspaceUserRole = getMemberRole(member);
+  const roles: WorkspaceUserRole[] = getMemberRoles(member);
   const memberIsAdmin = getIsAdmin(member);
   const { updateWorkspaceMemberRole } = hooks.useUpdateWorkspaceMemberRole();
-  const [selectedRole, setSelectedRole] = useState<WorkspaceUserRole>(role);
-
-  useEffect(() => setSelectedRole(role), [role]);
+  const [loading, setLoading] = useState(false);
 
-  const selectRole = (updated: WorkspaceUserRole) => {
-    onClick();
-    let roles: WorkspaceUserRole[];
-
-    if (updated !== "admin") {
-      roles = [updated];
-      if (memberIsAdmin) {
-        roles.push("admin");
-      }
-    } else {
-      roles = [role];
-      if (!memberIsAdmin) {
-        roles.push("admin");
-      }
+  const selectRole = (role: WorkspaceUserRole) => {
+    let updatedRoles = new Set(roles);
+    switch (role) {
+      case "viewer":
+        updatedRoles.delete("debugger");
+        break;
+      case "contributor":
+      case "debugger":
+        updatedRoles.add(role);
+        break;
+      case "admin":
+        updatedRoles.has(role) ? updatedRoles.delete(role) : updatedRoles.add(role);
+        break;
```

Comment:
The way this is written is confusing to me.

Why does "viewer" _delete_ a role from the set, but "debugger" _adds_ one? And why does "admin" _add or remove_ based on the role's presence? 

I think you're trying _not_ to delete one role when adding another, but this looks like an overly complex way to do that. Particularly the "admin" option (which is a checkbox) seems like it should be passing in a true/false flag.

Reply:
It is confusing. I'll do some refactoring to clarify what's happening.

Reply:
I pushed a refactoring commit with separate handlers for each option with comments. Hopefully that adds a bit of clarity.